File: cookblog-frontend/src/app/api/requests/update-recipe.request.ts
import { IngredientResource } from '../resources/ingredient.resource';
import { ImageResource } from '../resources/image.resource';
import { CategoryResource } from '../resources/category.resource';

export class UpdateRecipeRequest {
  private readonly _title: string;
  private readonly _description: string;
  private readonly _ingredients: IngredientResource[];
  private readonly _instructions: string;
  private readonly _image: ImageResource;
  private readonly _category: CategoryResource;
  private readonly _preparationTimeInMinutes?: number;
  private readonly _portions?: number;
  private readonly _calorificValue?: number;

  constructor(
    title: string,
    description: string,
    ingredients: IngredientResource[],
    instructions: string,
    image: ImageResource,
    category: CategoryResource,
    preparationTimeInMinutes?: number,
    portions?: number,
    calorificValue?: number,
  ) {
    this._title = title;
    this._description = description;
    this._ingredients = ingredients;
    this._instructions = instructions;
    this._image = image;
    this._category = category;
    this._preparationTimeInMinutes = preparationTimeInMinutes;
    this._portions = portions;
    this._calorificValue = calorificValue;
  }

  get title() {
    return this._title;
  }

  get description() {
    return this._description;
  }

  get ingredients() {
    return this._ingredients;
  }

  get instructions() {
    return this._instructions;
  }

  get image() {
    return this._image;
  }

  get category() {
    return this._category;
  }

  get preparationTimeInMinutes() {
    return this._preparationTimeInMinutes;
  }

  get portions() {
    return this._portions;
  }

  get calorificValue() {
    return this._calorificValue;
  }
}

File: cookblog-backend/src/main/java/com/example/cookblog/dto/requests/UpdateRecipeRequest.java
package com.example.cookblog.dto.requests;

import com.example.cookblog.dto.resources.CategoryResource;
import com.example.cookblog.dto.resources.ImageResource;
import com.example.cookblog.dto.resources.IngredientResource;
import com.fasterxml.jackson.annotation.JsonAutoDetect;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import lombok.Builder;
import lombok.extern.jackson.Jacksonized;

import java.util.Set;

@Builder
@Jacksonized
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public record UpdateRecipeRequest(
        @NotBlank @Size(max = 255) String title,
        @NotBlank @Size(max = 2000) String description,
        @NotNull Set<@NotNull @Valid IngredientResource> ingredients,
        @NotBlank @Size(max = 5000) String instructions,
        @NotNull @Valid ImageResource image,
        @Positive Integer preparationTimeInMinutes,
        @Positive Integer portions,
        @Positive Integer calorificValue,
        @NotNull @Valid CategoryResource category) {

}